Add tags to tests/suites (#3029)

jorge-vasquez-2301 opened the pull request:
Closes #3029

jorge-vasquez-2301 commented:
@jdegoes @adamgfraser I've added tags to some of the tests in `ZIOSpec`, and before I continue I would like to get some feedback from you guys, do you think the tags I'm creating/applying make sense? Maybe I'm creating too many categories or am I in the right path?

adamgfraser commented:
@jorge-vasquez-2301 First, I love what you did with the ZIO Tags! That is really nice.

My suggestion would be that we approach this incrementally. There are thousands of tests and a lot of different ways to categorize so I tend to think that a good PR would make incremental progress on this and not be a comprehensive solution.

What if we initially focus on a couple "high priority" tags and try to define those tags and tag everything related to them and then we can expand out to other takes or potentially merge with some tags done and pursue others separately.

I think if you look at the existing string tags so far (e.g. interruption, supervision, interop, future) they are probably a good starting point and plenty of work to do initially.

One other idea that I think might be good to explore with the ZIO Tags is some notion of sub typing. For example, right now we have "future" and "interop" tags. I would think "future" is a subset of "interop" in that "future" is one specific other protocol that we want to interop with. So it could be nice if we can express that so we can just tag a test with future and it will run whether we run all tests tagged future or interop.

Does that make sense?

jorge-vasquez-2301 commented:
@adamgfraser OK, I think I get the idea, I'm thinking about a solution like this for `ZIOTag`, what do you think?
```scala
  sealed trait ZIOTag {
    val value: String
    val subTags: List[ZIOTag] = Nil
  }
  object ZIOTag {
    case object ErrorHandling extends ZIOTag { override val value = "error-handling" }
    case object Future        extends ZIOTag { override val value = "future"         }
    case object Interop extends ZIOTag {
      override val value                 = "interop"
      override val subTags: List[ZIOTag] = List(Future)
    }
    case object Interruption extends ZIOTag { override val value = "interruption" }
    case object Regression   extends ZIOTag { override val value = "regression"   }
    case object Supervision  extends ZIOTag { override val value = "supervision"  }
  }

  def zioTag(zioTag: ZIOTag, zioTags: ZIOTag*): TestAspectPoly = {
    val tags = zioTags.map(_.value) ++ getSubTags(zioTag) ++ zioTags.flatMap(getSubTags)
    TestAspect.tag(zioTag.value, tags.distinct: _*)
  }

  private def getSubTags(zioTag: ZIOTag): List[String] = {
    @tailrec
    def loop(currentZioTag: ZIOTag, remainingZioTags: List[ZIOTag], result: List[String]): List[String] =
      (currentZioTag.subTags, remainingZioTags) match {
        case (Nil, Nil)      => currentZioTag.value :: result
        case (Nil, t :: ts)  => loop(t, ts, currentZioTag.value :: result)
        case (st :: sts, ts) => loop(st, sts ++ ts, currentZioTag.value :: result)
      }
    zioTag.subTags match {
      case t :: ts => loop(t, ts, Nil)
      case Nil     => Nil
    }
  }
```

jdegoes commented on core-tests/shared/src/test/scala/zio/ZIOBaseSpec.scala:
```diff
@@ -5,4 +5,36 @@ import zio.test._
+    val value: String
+  }
+  object ZIOTag {
+    case object Caching extends ZIOTag { override val value = "caching" }
```
From the perspective of a user who will be running these, I suggest each tag be 1 word long, maximum. And shortest form of a word, e.g. `ErrorHandling` => `Errors`, `Memoization` => `Memo`, `ResourceHandling` => `Resource`, etc.

jorge-vasquez-2301 replied:
Yeah, makes sense, thanks for the suggestion @jdegoes!

adamgfraser commented:
@jorge-vasquez-2301 Yes. I think you could simplify by using a case class.

jorge-vasquez-2301 commented on core-tests/shared/src/test/scala/zio/ZIOBaseSpec.scala:
```diff
@@ -1,8 +1,45 @@
+  }
+  object ZIOTag {
+    case object Errors        extends ZIOTag { override val value = "errors" }
+    case object FutureInterop extends ZIOTag { override val value = "future" }
```
I had to use two words for this tag: `FutureInterop`, I couldn't make it just `Future` because it would clash with scala `Future`, do you have any ideas for making this just one word @adamgfraser @jdegoes?

jorge-vasquez-2301 commented:
@jdegoes @adamgfraser could you please take a look now to this PR? I've added tags to several tests, and want to know your opinion and if the tags I've added are sufficient for now

jorge-vasquez-2301 commented:
@jdegoes Just updated the PR according to your suggestions 🙂

jdegoes commented:
@jorge-vasquez-2301 Thank you for your contribution!

jorge-vasquez-2301 commented:
@jdegoes @adamgfraser I'm happy to help! 😄
